### ISSUE TITLE
Issues in qiskit-alt a Julia backend for Qiskit.

### ISSUE THREAD

**Issue by jlapeyre** [qiskit-advocate/qamp-spring-22#16]
### Description

[qiskit-alt](https://github.com/Qiskit-Extensions/qiskit-alt) is a high-performance Julia backend for Qiskit with a Python frontend. It is quite new and experimental. It is not distributed as part of Qiskit. The main application is computing a qubit Hamiltonian from the specification of an electronic structure problem.

There are a number of projects available, small and large. The appropriate project depends on the experience of the mentee. There are a number of issues in the qiskit-alt repo. But, the project is divided into several other repos, most of which are [hosted here](https://github.com/Qiskit-Extensions). These repos have their own issues. A major part of the effort is improving deployment options of Julia with the Python project. These issues are in [`julia_project`](https://github.com/jlapeyre/julia_project)

Note that much of the work happens on the Python side. So, if you know some Python, but don't yet know Julia, you can still contribute.

### Deliverables

Open a pull request that closes an issue and work through getting it merged. Alternatively, start by opening a new issue; for instance as a sub-issue of a more open ended one.

### Mentors details

- Mentor 1
  - Name: John Lapeyre
  - GitHub ID: @jlapeyre
  - What they do: Qiskit development
- Mentor 2
  - Name: Jim Garrison
  - GitHub ID: @garrison
  - What they do: Qiskit applications development

### Number of mentees

2

### Type of mentees

- Mentee *
  - Required:
    - Some coding and github skills.
    - Desire to contribute to an opensource project.
  - Nice to have:
    - Python experience
    - Julia experience
    - Quantum computing knowledge

**Comment by gunchamalik** [comment]
@jlapeyre I have basic knowledge pf python, but no knowledge of Julia. I am interested in learning and contributing to this project.

**Comment by jlapeyre** [comment]
Hi @gunchamalik, thanks for reaching out! We will contact you very soon.

**Comment by jlapeyre** [comment]
Here is one possible topic:

There are two separate ways for Julia and Python to communicate dynamically. One is via [PyCall.jl](https://github.com/JuliaPy/PyCall.jl) / [pyjulia](https://github.com/JuliaPy/pyjulia). The other (more recent) is via [PythonCall.jl / juliacall](https://github.com/cjdoris/PythonCall.jl). It would be great if a single candidate were the clear winner. It's not clear that one way is better than the other; they do have different features.

We need to find out whether to support one way or the other, or both, and if so how.  A good bit of infrastructure for investigating this is in place. (I need to push some changes today or tomorrow to make this visible)  

I'll refer to the communication methods as `PyCall.jl` and `PythonCall.jl` which are the Julia modules, but the corresponding Python modules are implied.

Some possibilities are

* Support both `PyCall.jl` and `PythonCall.jl`  as backends to a single user-facing API. This is probably not a good idea. The semantics of the two systems are sufficiently different to require a fair amount of complexity to handle this. All future additions to the Julia code in qiskit_alt would have to support this.
* Support both `PyCall.jl` and `PythonCall.jl`, but in parallel. That is, some code would use one backend and some the other. This might be confusing because the semantics and coding styles might be different and cause lack of clarity in the code base. The user need not see this because it shouldn't be difficult to provide a uniform pure-python interface.
However, more importantly, the two systems handle wrapping and converting data differently. So they can't always interact transparently without yet more conversion.
* Support only one of the two` PyCall.jl` or `PythonCall.jl`. This requires some implementation and benchmarking (and thinking!) to inform the decision for one system or the other.

I will follow up here with a link to the latest version of qiskit_alt, which is set up for investigating the topics above.

**Comment by rafal-pracht** [comment]
[# 16 Julia - Qiskit.pdf](https://github.com/qiskit-advocate/qamp-spring-22/files/8425516/16.Julia.-.Qiskit.pdf)

**Comment by gunchamalik** [comment]
[Julia Backend for Qiskit - Checkpoint 1.pdf](https://github.com/qiskit-advocate/qamp-spring-22/files/8425788/Julia.Backend.for.Qiskit.-.Checkpoint.1.pdf)

**Comment by saharbenrached** [comment]
[#16 Issues in qiskit-alt: a Julia backend for Qiskit.pdf](https://github.com/qiskit-advocate/qamp-spring-22/files/8429107/16.Issues.in.qiskit-alt.a.Julia.backend.for.Qiskit.pdf)

**Comment by MozammilQ** [comment]
This is my checkpoint-1's small presentation:
[#16 Issues in qiskit-alt: a Julia backend for Qiskit.pdf](https://github.com/qiskit-advocate/qamp-spring-22/files/8438884/16.Issues.in.qiskit-alt.a.Julia.backend.for.Qiskit.pdf)

**Comment by rafal-pracht** [comment]
The work on Julia-powered quantum library QuantumCircuits (https://github.com/Adgnitio/QuantumCircuits.jl) is going very well.
![image](https://user-images.githubusercontent.com/55279376/166814772-5dd221ea-b492-4be0-a60b-9f91dd0ee3ce.png)

There is now 5 Contributor in the library who created and merged 24 pull requests and by that closed 21 issues (https://github.com/Adgnitio/QuantumCircuits.jl/issues?q=is%3Aissue+is%3Aclosed). 
![image](https://user-images.githubusercontent.com/55279376/166814807-7508991f-f465-494f-b088-251edcd1300b.png)


During these works, we focused on fixing the performance issue and did some code cleaning:
![image](https://user-images.githubusercontent.com/55279376/166814897-1c077cba-48a5-4d96-a310-ca08874a0fba.png)


- We changed the code so that we don't do the unnecessary memory allocation. 
- We used the cis method instead of exp for the complex number exponential function and by that, we gain on performance.
- We deleted unneeded functions such as toString and dagger.
- We changed the types of functions arguments to a more generic way.

Four issues were marked as "good first issues" so anyone who would like to help us is more than welcome.

But there is a lot of the work to do. We have to move from Qiskit-style circuit creation to a more Julia idiomatic way that uses the macro to create domain-specific language dedicated to quantum operation. The other thing we should do is avoid converting the circuit to a Qiskit object during the draw procedure. So there is a lot of task before us :)

**Comment by saharbenrached** [comment]
The modue qiskit-alt can be used to compute the fermionic operator of a molecule and convert it to a Pauli operator using the Jordan-Wigner transformation. This would later serve as an input to run molecular simulation algorithms from Qiskit Nature module. 
We are adding tutorials to demonstrate the use of the qiskit-alt module for users to run molecular simulation tasks. 
For the first tutorial, we are running the Variational Quantum Eigensolver to simulate the ground state energy of the Hygrogen molecule. In the first part, instead of using qiskit-nature, we use qiskit-alt features to describe the Fermionic operator and convert it to a Pauli operator with the Jordan-Wigner transform. The operator is then mapped to a quantum circuit and will be the input to the VQE method from Qiskit.
![molecule-description](https://user-images.githubusercontent.com/58570811/166969708-cdd44fac-5e8e-4f0e-bb0d-2e00b97e2bd7.png)
We compute the ground state energy of the Hydrogen molecule
![running-VQE](https://user-images.githubusercontent.com/58570811/166969780-090e1ba6-4f7b-412b-8e5a-0cc9716da21b.png)
However, the output result is not precise. 
Our main goal for the next stage of this project is to improve the computation accuracy.  We are still investigating at which level the error is caused. We will continue fixing issues and creating more tutorials for the simulation of more complex molecule models (if possible) and other chemistry simulation algorithms such as computing higher state energies.

**Comment by MozammilQ** [comment]
After Checkpoint 1 I focused on the issue: https://github.com/Qiskit-Extensions/qiskit-alt/issues/27

In qiskit-alt/docker_tests, there is a shell file ./run_dockerfiles.sh
Basically, this is a shell script that has purposes like building a docker image, running a docker container, and then running different tests for qiskit-alt package, other functionality includes like logging into the shell, etc in all these most of the variables like image name of the docker container its tag and other information was hardcoded along with this the functionality/job/behavior of the script depends on the command line argument. Implementing this group of functionality with this set of variables was good enough for a small scale but these were not easy to use, and, were not scalable to more number of other functionality.

I implemented a Python Command Line Tool which would use its command-line argument to take different variables like names of the docker images, its tag name, etc along with the behavior/functionality to execute. The good thing about this was the scalable numbers of variables and functionalities/behavior that can be implemented with the neatness of Python.
The CLI Tool is named ./run_dockerfile.py and it exists in the same directory as its shell counterpart.

```shell
./run_dockerfile.py –help
```
gets us this output:
![image_1](https://user-images.githubusercontent.com/54737489/167042043-b4f5ce34-79a6-4de7-9faf-b7a0fdd13848.png)

As an example 
```shell
./run_dockerfile.py –action=run
```
gets us this output:
![image_2](https://user-images.githubusercontent.com/54737489/167042284-fb228fbc-b788-4f81-932a-93508f03c2d4.png)

Here is the Pull Request: https://github.com/Qiskit-Extensions/qiskit-alt/pull/30



Just during the checkpoint 1, I was working on this issue: https://github.com/Qiskit-Extensions/qiskit-alt/issues/21

After, 
```shell
pip install qiskit-alt
```

The Python Code:
```python
import qiskit_alt
qiskit_alt.project.ensure_init()
```
should provide us with julia installed and qiskit_alt ready to accept commands but this terminated with this error:
![image_3](https://user-images.githubusercontent.com/54737489/167042362-1ed12b7c-551e-46de-bd0c-aa76785f1dd3.png)


The issue was in a python package called find_julia: https://github.com/jlapeyre/find_julia
It’s purpose is to make sure julia is present if not then install it.

After the issue was fixed, the output looked like this:
![image_4](https://user-images.githubusercontent.com/54737489/167042380-cec53679-e413-4d54-a65c-f217439d6bc1.png)


This was the Pull Request: https://github.com/jlapeyre/find_julia/pull/3

**Comment by rafal-pracht** [comment]
[Checkpoint3-#16.pdf](https://github.com/qiskit-advocate/qamp-spring-22/files/8869398/Checkpoint3-.16.pdf)

**Comment by MozammilQ** [comment]
Final Presentation:
https://youtu.be/bjz3FKpm744
